### ISSUE TITLE
Inconsistent rate of receiving messages

### ISSUE THREAD

**Issue by vulkoingim** [GoogleCloudPlatform/google-cloud-go#949]
## Client
PubSub - `4b98a6370e36d7a85192e7bad08a4ebd82eac2a8`
## Describe Your Environment
Go 1.10
Debian 7
## Expected Behavior
Messages are consumed fast.
## Actual Behavior
Messages arrive really slowly.

We are experiencing a very weird issue with PubSub, where seemingly at random we experience a delay in processing the messages. The delay does not correspond with the amount of messages, actually it seems that the more messages there are it runs smoother. We are on the `4b98a6370e36d7a85192e7bad08a4ebd82eac2a8` commit of the go library. We don't encounter errors of any kind while pulling or processing the messages.

A bit about our client: 

It essentially pulls from N (3 at this moment) subscriptions, each in a separate goroutine. Each goroutine itself spawns X (5 right now ) more goroutines to consume the messages. The code looks something like: 

```
ctx := context.Background()

for _, s := range subscriptions {
        go func(this *Subscription) {
                toAck := make(chan *pubsub.Message)
                acks(ctx, toAck)
                
                erPull := this.Receive(ctx, func(lctx context.Context, msg *pubsub.Message) {
                        toAck <- msg
                })
        }(s)
}

func acks(ctx context.Context, toAck chan *pubsub.Message) {
	for i := 0; i < 5; i++ {
		go func(lctx context.Context) {
			for {
				select {
				case <-lctx.Done():
					return
				case msg := <-toAck:
                                        ...process...
					msg.Ack()
				}
			}
		}(ctx)
	}
}
```

The processing of a message happens very fast, < 1ms, with some outliers taking 2-3ms more. 

We have experimented with different subscription ReceiveSettings, it doesn't seem to have any effect (except processing messages slower when there is a burst and the gouroutnes and MaxOutstandingMessages are set to low values). 

Current:
```
MaxExtension:           60 * time.Second,
MaxOutstandingMessages: -1,
MaxOutstandingBytes:    2e9,
NumGoroutines:          5,
```

This code runs on 5 servers, with plenty of resources available. We have bursts of ~2k/s at most on one of the subscriptions, while the rest usually get less, which really is nothing. We've tested with 20k+/s (from cold start, so it's not the burst causing the issue) and it consumes them flawlessly. We have metrics for everything, and have added opencensus tracing as well. The opencensus metrics available are (we have them only since yesterday): 

```
ocgrpc.ClientRequestCountView
ocgrpc.ClientErrorCountView
ocgrpc.ClientRequestBytesView
ocgrpc.ClientResponseBytesView
ocgrpc.ClientRoundTripLatencyView
pubsub.PullCountView
pubsub.AckCountView
pubsub.NackCountView
pubsub.ModAckCountView
pubsub.StreamOpenCountView
pubsub.StreamRetryCountView
pubsub.StreamRequestCountView
pubsub.StreamResponseCountView
```
The following metrics are related only to a single topic/subscription

StackDriver Topic:

![stackdriver-topic-5 15am](https://user-images.githubusercontent.com/5501705/38028080-103f9dd2-3292-11e8-9578-cdbece95676d.png)

StackDriver Subscription:

![stackdriver-subs-5 15am](https://user-images.githubusercontent.com/5501705/38028083-130ff6a6-3292-11e8-8173-e30357a31750.png)

OpenCensus:

![opencensus-5 15am](https://user-images.githubusercontent.com/5501705/38028099-1bc74c5e-3292-11e8-8dd6-633ebfe3fb40.png)

Prometheus internal[3m]:

![prometheus-5 15](https://user-images.githubusercontent.com/5501705/38028115-28318180-3292-11e8-968f-172cceec1570.png)

Prometheus internal[1m]:

![prometheus-5 15am-1m](https://user-images.githubusercontent.com/5501705/38028239-a2a5eaf0-3292-11e8-8c96-d081b64d2c9f.png)

Prometheus metrics from processing larger amount of messages without delay (~11:23) + Delay when processing almost no messages (~10:40)

![prometheus-11 23](https://user-images.githubusercontent.com/5501705/38029436-2414ee76-3296-11e8-8f67-2c96e33f6587.png)

The whole issue is super difficult to debug, so far we have been unable to reproduce it synthetically and we can't seem to find a pattern for when it occurs...

**Comment by jba** [comment]
Thanks for the detailed report. 

Where are the receiving processes running? GCE? GKE?

**Comment by vulkoingim** [comment]
They are running outside of GCP. I will provide you with metrics from within GCE as well in the coming day(s).

**Comment by vulkoingim** [comment]
I already had those metrics, although not with the extra code you have there. Anyway, here there are for the same timeframe I posted above...

Acks:
![acks](https://user-images.githubusercontent.com/5501705/38087630-17b0120a-3359-11e8-902c-6ed87b9569fc.png)

Retries:
![retry](https://user-images.githubusercontent.com/5501705/38087633-1e934c9a-3359-11e8-8196-c0eea95aaf9f.png)


Stream open:
![stream_open](https://user-images.githubusercontent.com/5501705/38087640-23d4ec36-3359-11e8-8dff-5b4a61a1a2de.png)


I've highlighted the subscription with the large delay, which is the same one as the screenshots above.

Also, since last night all receivers are running on GCE VM's and the issue persists.

**Comment by jba** [comment]
Let me see if I can summarize these graphs. After each description I refer to the relevant graph by name.

- For about two minutes, from around 5:14 to 5:16, there was a burst of published messages, peaking  at around 600/s. I'd estimate about 50,000 messages in all (StackDriver Topic, Publish Message Operations).

- The receiving program is acking messages at a high rate, with a peak of about 830/s (OpenCensus, ack_count). I estimate that you should be able to ack well over 1000 messages/s (a conservative 5ms/msg * 5 goroutines) so it doesn't look like your program is being saturated. 

- The stream remained open for the entire time (OpenCensus, stream open). 

- During the peak, a backlog builds up of 2,500 undelivered messages (around 5%, by my estimate). It gradually resolves after about a minute (StackDriver Subscription, Undelivered Messages).

- By about 5:17, 5% of the messages had taken more than 43s to process (Prometheus internal[1m], Time since Publish). What is the exact definition of Time to Publish, by the way?

- In a separate incident (third Prometheus graph), one or two messages took 25s to process, during a period of low activity. We don't have client-side data (ack count, stream opens) about this incident.

I don't have any thoughts about what's causing this, but I wanted to make sure we agreed on what the problem is.

I'm out for a week starting today, but @jadekler and others will pick this up next week.

**Comment by jadekler** [comment]
I'd like to know how you're calculating `Time since Published`? That seems to be the key metric here (correct me if I'm wrong, I don't see anything else above indicating slowness).

**Comment by vulkoingim** [comment]
Regarding jba's summaries: that's pretty much what is happening. As to your last point, here are the opencensus metrics during the last graph's period:

Acks:
![ack_count_28 03](https://user-images.githubusercontent.com/5501705/38240229-5746cfc4-372f-11e8-8815-5d3e21feb3ab.png)

Stream open/retry:
![stream_retry_open_28 03](https://user-images.githubusercontent.com/5501705/38240239-5f9977a8-372f-11e8-88f7-20f262dd014a.png)


@jadekler `Time since Published` is basically the time since the message got published, processed and ack'd, i.e. `time.Since(pubsub.Message.PublishTime)`. Or to illustrate it in the code I posted above:

func acks(ctx context.Context, toAck chan *pubsub.Message) {
	for i := 0; i < 5; i++ {
		go func(lctx context.Context) {
			for {
				select {
				case <-lctx.Done():
					return
				case msg := <-toAck:
                                          prcessingTimeStart := time.Now()

                                        ...process...
					msg.Ack()
                                          
                                          // Beaniator Processing Time
                                         prcessTime.Observe(time.Since(processingTimeStart).Seconds())

                                          // Time Since Publish
                                         timeSincePublish.Observe(time.Since(msg.PublishTime).Seconds())
				}
			}
		}(ctx)
	}
}

**Comment by jadekler** [comment]
@vulkoingim I'm a bit slow to understand this issue, so please correct me if I misunderstand anything here. This seems to be the problem statement:

> We are experiencing a very weird issue with PubSub, where seemingly at random we experience a delay in processing the messages.

Processing the message is time between arrival of message and right after ack. I don't see a graph of this in the above - am I missing it?

Or more broadly, could you point me to what makes you believe that process time is slower than you'd expect, or inconsistent?

**Comment by vulkoingim** [comment]
@jadekler poor wording on my part. We are experiencing a delay in receiving the messages. What makes me believe that it is slower and inconsistent.... If you look at the last 2 graphs of my original post you can see that a similar volume results in different `Time since publish`. The processing time is almost the same in both cases. 

As to the time `between receiving and ack`, you can see it on the graph `Beaniator processing time`. True, this is the time since the message is received on the channel, and not since `subscription.Receive()`, but there is no substantial difference between the two. I don't have metrics for it during the period of the graphs I posted above, but here you can see it in an incident from yesterday (we have metrics for it as well since a few days ago):

![screen shot 2018-04-04 at 10 38 54](https://user-images.githubusercontent.com/5501705/38298028-6da028f2-37f6-11e8-87a4-d4581e4ac46b.png)

In other words, yes, I am interested in the time between ` publishing (and processing) and right after ack`. 

Processing time, defined as: `time between arrival of message and right after ack` is not slower, therefore I am lead to believe that there is something fishy either on PubSub side, or on the go client which prevents the application from pulling available messages at a high(consistent) rate.

**Comment by jadekler** [comment]
I think I get what's going on. Tell me if any of this is wrong:

- You have a consumer running, waiting for things to arrive
- A bunch of messages get published in a short amount of time
- The consumer begins chugging through the messages with very little time between receive and ack
- You note that 50% take 30s, 95% take 1.02m, 99% take 1.08m

So I think if you plotted the time it takes to process against the order in which the message was published, you would see a uniformly increasing ladder. Something like this:

![screen shot 2018-04-04 at 4 56 03 pm](https://user-images.githubusercontent.com/3584893/38340589-156147ce-3829-11e8-8e86-c09701d5c372.png)

I ran a demo where I published 1,000,000 ~100byte messages quickly while a consumer was reading. The above is the microseconds between time.Now() and msg.PublishTime. The notable feature is that `publishing (and processing) and right after ack` time increases linearly with respect to order in which the message was published. Basically the app doesn't get all the outstanding messages at once. Here are two reasons why, although many more exist:

- grpc message limit is 4mb, so things above the limit have to wait (and that goes both ways - reception of large messages, _and_ sending a bunch of acks / modacks back at the server)
- Buffers exist not just at the client level but also in all network devices between you and pubsub server

I wonder if that by itself explains your situation? Have you tried horizontally scaling? If you're trying to handle burst-y load, I _suspect_ that you'll have better luck scaling out than scaling up.

Postnote: I made a lot of assumptions about your situation above - please feel free to correct me on any of those.

**Comment by vulkoingim** [comment]
That is not the case. I understand if there are way too many messages than the consumer is able to handle, there will be delays. 

A single consumer is able to process more than 2k/s. The consumer is already scaled horizontally (at this moment there are 3 of them running), that is definitely not the bottleneck.

I am well aware of all those limitations. The whole reason I opened this ticket is because none of this makes sense. As I mentioned and as you can see on the graphs above this doesn't happen when the volume is too high for the consumer to be able to process. In fact it seems that there are less issues when there are more messages available for processing. 

Here is another fresh example from yesterday. I have highlighted the time with the higher volume on both graphs: ~17:21. At that time 99% of messages were `published, processed, acked` within 6ms. With messages being published to the topic at 1300/s


Around ~14:31 messages were being published to the topic with a rate of ~750/s and  99% of `publish, process,ack` completed in ~2m 15s


![screen shot 2018-04-05 at 10 37 40](https://user-images.githubusercontent.com/5501705/38356151-43652334-38bf-11e8-8a6f-b4592ad14692.png)
![screen shot 2018-04-05 at 10 38 09](https://user-images.githubusercontent.com/5501705/38356155-45d411ca-38bf-11e8-9296-5d05fb1ee5ae.png)

**Comment by jadekler** [comment]
Ahh ok. I need to think about this a bit more and maybe see if I can repro your situation more. Could you tell me about your publisher? Is it written in Go, and is there only one publisher (or, many)?

**Comment by vulkoingim** [comment]
Glad to hear that it doesn't make sense not only to me :)

The publishers' code is written in PHP and there are many of them. The settings are default and are using GRPC.

**Comment by jadekler** [comment]
Thanks @vulkoingim. I'm going to spend time today trying to reproduce this issue. Appreciate your patience.

**Comment by jadekler** [comment]
Update to this - I was (unfortunately? fortunately?) able to reproduce this issue using the following programs:

Receiver:
```
package main

import (
	"cloud.google.com/go/pubsub"
	"context"
	"fmt"
	"time"
)

const (
	projectID      = "some-sandbox"
	topicID        = "php-test-topic"
	subscriptionID = "some-go-to-php-subscription"
)

func main() {
	fmt.Println("Go starting up")
	ctx := context.Background()

	client, err := pubsub.NewClient(ctx, projectID)
	if err != nil {
		panic(err)
	}

	t := client.Topic(topicID)
	exists, err := t.Exists(ctx)
	if err != nil {
		panic(err)
	}
	if !exists {
		t, err = client.CreateTopic(ctx, topicID)
		if err != nil {
			panic(err)
		}
	}

	_, err = client.CreateSubscription(ctx, subscriptionID, pubsub.SubscriptionConfig{Topic: t})
	if err != nil {
		fmt.Println(err)
		fmt.Println("Go tried to create a subscription but it already exists; continuing on")
	}

	sub := client.Subscription(subscriptionID)
	exists, err = sub.Exists(ctx)
	if err != nil {
		panic(err)
	}
	if !exists {
		sub, err = client.CreateSubscription(ctx, subscriptionID, pubsub.SubscriptionConfig{Topic: t})
		if err != nil {
			panic(err)
		}
	}

	sub.Receive(ctx, func(c context.Context, m *pubsub.Message) {
		fmt.Println(time.Now().Sub(m.PublishTime).Nanoseconds() / 1000000, "ms")
		m.Ack()
	})
}
```

Publisher:
```
<?php

# Includes the autoloader for libraries installed with composer
require __DIR__ . '/vendor/autoload.php';

# Imports the Google Cloud client library
use Google\Cloud\PubSub\PubSubClient;

# Your Google Cloud Platform project ID
$projectId = 'some-sandbox';

# Instantiates a client
$pubsub = new PubSubClient([
    'projectId' => $projectId
]);

# The name for the new topic
$topicName = 'php-test-topic';

// Get an instance of a previously created topic.
$topic = $pubsub->topic($topicName);

# Creates the new topic
//$topic = $pubsub->createTopic($topicName);
//(already exists)
//echo 'Topic ' . $topic->name() . ' created.';

for ($j = 1; $j <= 10000; $j++) {
    for ($i = 1; $i <= rand(0,1000); $i++) {
        // Publish a message to the topic.
        $topic->publish([
            'data' => "My new message. $j-$i",
            'attributes' => [
                'location' => 'Detroit'
            ]
        ]);
        echo "published $j-$i\n";
    }

    sleep(rand(0, 1000));
}
```

![screen shot 2018-04-10 at 9 47 59 am](https://user-images.githubusercontent.com/3584893/38571078-5ace9094-3ca4-11e8-8de8-7f1673df422f.png)

@vulkoingim Next steps for me are to see if this is a php-publisher specific issue (very unlikely), and discuss internally with the pubsub team to see what they think may cause this. I'm not aware of any client library issue that would cause this.

**Comment by jba** [comment]
@jadekler Definitely fortunately!

**Comment by jadekler** [comment]
Hi @vulkoingim.

Apologies for the long delay. I've spent time testing and we believe we understand this problem well. We're digging a bit deeper for a more complete understanding and a fix for this. I'll report back with more details once that's done.

Thanks again for reporting this.

**Comment by jba** [comment]
We think we've fixed this, but can't really test. Please verify.

**Comment by jadekler** [comment]
Just to add to the discussion, my repro no longer shows any spikes / stuff above 1500ms:

<img width="593" alt="screen shot 2018-04-23 at 9 31 17 am" src="https://user-images.githubusercontent.com/3584893/39140415-5f758b72-46d9-11e8-8392-94c621fd4659.png">
